src/web/api/badges.rs

```diff
@@ -0,0 +1,337 @@
+
+        (rows, version.to_owned())
+
+    // If no version was provided, find the latest release and use that
+    } else {
+        let query = "
+            SELECT is_library, rustdoc_status, build_status, yanked, version
+            FROM releases
+            WHERE id IN (
+                SELECT latest_version_id from crates
+                WHERE name = $1
+            )";
+        let rows = ctry!(conn.query(query, &[&krate]));
+
+        let version = api_error!(
+            rows.iter().next().map(|r| r.get::<_, String>("version")),
+            "The requested crate does not exist",
+        );
+
+        (rows, version)
+    };
+
+    // If the crate & version is found in the database, it's built in some form
+    let build_status = if let Some(release) = rows.iter().next() {
+        // If the release isn't a library
+        if !release.get::<_, bool>("is_library") {
+            BuildStatus::NotLibrary
+
+        // If the release was yanked
+        } else if release.get("yanked") {
+            BuildStatus::Yanked
+
+        // If the build succeeded
+        } else if release.get("rustdoc_status") || release.get("build_status") {
+            BuildStatus::Built
+
+        // If none of the above, then the build failed in some way
+        } else {
+            BuildStatus::Failed
+        }
+
+    // If we can't find the crate & version in the db, it might be in the queue
+    } else {
+        let query = "SELECT COUNT(*) AS count FROM queue WHERE name = $1 AND version = $2";
+        let count: i64 = ctry!(conn.query(query, &[&krate, &version]))
+            .iter()
+            .next()
+            .map(|r| r.get("count"))
+            .unwrap_or_default();
+
+        // If there's an entry for the crate in the build queue, it's queued
+        if count != 0 {
+            BuildStatus::Queued
+        } else {
+            return ApiErrorV1::new("The requested crate does not exist").into_response();
+        }
+    };
+
+    // Form the url of the crate
+    let docsrs_url = format!("https://docs.rs/crate/{}/{}", krate, version);
+
+    let badge = BadgeV1 {
+        name: krate.to_owned(),
+        version,
+        docsrs_url,
+        build_status,
+    };
+
+    let mut resp = Response::with((status::Ok, serde_json::to_string(&badge).unwrap()));
+    resp.headers
+        .set(ContentType("application/json".parse().unwrap()));
+
+    Ok(resp)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test::wrapper;
+
+    #[test]
+    fn get_latest() {
+        wrapper(|env| {
+            let web = env.frontend();
+            let db = env.db();
+
+            db.fake_release()
+                .name("saltwater")
+                .version("1.4.0")
+                .create()?;
+
+            let mut resp = web.get("/api/v1/badges?crate=saltwater").send()?;
+
+            assert!(resp.status().is_success());
+            assert_eq!(
+                resp.json::<BadgeV1>()?,
+                BadgeV1 {
```

Comment:
When doing tests on the API I'd match with `serde_json::Value` (generating the expected result with the `json!` macro), so we can guarantee the result will be the same regardless of bugs or changes in the serialization code.